File: runtime/native/osx/src/TiWindowConfig.m
/**
 * This file is part of Appcelerator's Titanium project.
 *
 * Copyright 2008 Appcelerator, Inc.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *    http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License. 
 */
#import <WebKit/WebKit.h>
#import "TiWindowConfig.h"
#import "AppProtocol.h"

@implementation TiWindowConfig


- (id) init
{
	self = [super init];

	// setup defaults
	x=-1; // indicate not set
	y=-1; // indicate not set
	width=100;
	height=100;
	maxHeight=9000;
	maxWidth=9000;
	minHeight=0;
	minWidth=0;
	title=@"Titanium Application";
	url=@"index.html";
	chrome=NO;
	transparency=1.0;
	visible=YES;
	minimizable=YES;
	maximizable=YES;
	resizable=YES;
	fullscreen=NO;
	closeable=YES;
	scrollbars=YES;
	
	TRACE(@"TiWindowConfig: alloc\n");

	return self;
}

- (void)dealloc
{
	TRACE(@"TiWindowConfig: dealloc\n");
	window = nil;
	[title release];
	title = nil;
	[url release];
	url = nil;
	[super dealloc];
}

- (NSString *)description 
{
	return @"[TiWindowConfig native]";
}


- (NSString*) getID
{
	return windowID;
}

- (void) setID:(NSString*)i
{
	[windowID release];
	windowID = [i copy];
}


- (CGFloat) getX
{
	return x;
}

- (void) setX:(CGFloat)newx
{
	x = newx;
}

- (CGFloat) getY
{
	return y;
}

- (void) setY:(CGFloat)newy
{
	y = newy;
}

- (CGFloat) getWidth
{
	return width;
}

- (void) setWidth:(CGFloat)w
{
	width = w;
}

- (CGFloat) getHeight
{
	return height;
}

- (void) setHeight:(CGFloat)h
{
	height = h;
}

- (CGFloat) getMinWidth
{
	return minWidth;
}

- (void) setMinWidth:(CGFloat)w
{
	minWidth = w;
}

- (CGFloat) getMinHeight
{
	return minHeight;
}

- (void) setMinHeight:(CGFloat)h
{
	minHeight = h;
}

- (CGFloat) getMaxWidth
{
	return maxWidth;
}

- (void) setMaxWidth:(CGFloat)w
{
	maxWidth = w;
}

- (CGFloat) getMaxHeight
{
	return maxHeight;
}

- (void) setMaxHeight:(CGFloat)h
{
	maxHeight = h;
}

- (BOOL) isChrome
{
	return chrome;
}

- (void) setChrome:(BOOL)yn
{
	chrome = yn;
}

- (CGFloat) getTransparency
{
	return transparency;
}

- (void) setTransparency:(CGFloat)f;
{
	transparency = f;
}

- (BOOL) isVisible
{
	return visible;
}

- (void) setVisible:(BOOL)yn
{
	visible = yn;
}

- (BOOL) isMinimizable
{
	return minimizable;
}

- (void) setMinimizable:(BOOL)yn
{
	minimizable = yn;
}

- (BOOL) isMaximizable
{
	return maximizable;
}

- (void) setMaximizable:(BOOL)yn
{
	maximizable = yn;
}

- (BOOL) isResizable
{
	return resizable;
}

- (void) setResizable:(BOOL)yn
{
	resizable = yn;
}

- (BOOL) isFullscreen
{
	return fullscreen;
}

- (void) setFullscreen:(BOOL)yn
{
	fullscreen = yn;
}

- (BOOL) isCloseable
{
	return closeable;
}

- (void) setCloseable:(BOOL)yn
{
	closeable = yn;
}

- (BOOL) isScrollbars
{
	return scrollbars;
}

- (void) setScrollbars:(BOOL)yn
{
	scrollbars = yn;
}


- (NSString*) getTitle
{
	return title;
}

- (void) setTitle:(NSString*)t
{
	[title release];
	title = [t copy];
	
	if (window!=nil)
	{
		[window setTitle:title];
	}
}

- (NSString*) getURL
{
	return url;
}

- (void) setURL:(NSString*)u
{
	[url release];
	url = [u copy];
}

- (NSUInteger) toWindowMask
{
	NSUInteger mask = 0;
	if (chrome)
	{
		mask = NSBorderlessWindowMask;
	}
	else
	{
		mask |= NSTitledWindowMask;
		if (resizable)
		{
			mask |= NSResizableWindowMask;
		}
		if (closeable)
		{
			mask |= NSClosableWindowMask;
		}
		if (maximizable)
		{
			mask |= NSMiniaturizableWindowMask;
		}
		if (maximizable)
		{
			// handled in the window code
		}
	}
	return mask;
}

- (BOOL)urlMatches:(NSString*)testURL
{
	// first see if we match the full URL
	if ([url isEqualToString:testURL])
	{
		return YES;
	}
	NSString *a = [AppProtocol getPath:[NSURL URLWithString:url]];
	NSString *b = [AppProtocol getPath:[NSURL URLWithString:testURL]];
	TRACE(@"a=%@, b=%@, matches=%d",a,b,[a isEqualToString:b]);
	return [a isEqualToString:b];
}

- (void)assign:(TiWindow*)w
{
	window = w;
}

#pragma mark -
#pragma mark WebScripting

+ (BOOL)isSelectorExcludedFromWebScript:(SEL)sel {
	return (nil == [self webScriptNameForSelector:sel]);
}


+ (NSString *)webScriptNameForSelector:(SEL)sel 
{
	if (sel == @selector(getX)) {
		return @"getX";
	} else if (sel == @selector(setX:)) {
		return @"setX";
	} else if (sel == @selector(getY)) {
		return @"getY";
	} else if (sel == @selector(setY:)) {
		return @"setY";
	} else if (sel == @selector(getID)) {
		return @"getID";
	} else if (sel == @selector(getWidth)) {
		return @"getWidth";
	} else if (sel == @selector(setWidth:)) {
		return @"setWidth";
	} else if (sel == @selector(getHeight)) {
		return @"getHeight";
	} else if (sel == @selector(setHeight:)) {
		return @"setHeight";
	} else if (sel == @selector(setFullscreen:)) {
		return @"setFullscreen";
	} else if (sel == @selector(isFullscreen)) {
		return @"isFullscreen";
	} else if (sel == @selector(isResizable)) {
		return @"isResizable";
	} else if (sel == @selector(setResizable:)) {
		return @"setResizable";
	} else if (sel == @selector(isMaximizable)) {
		return @"isMaximizable";
	} else if (sel == @selector(setMaximizable:)) {
		return @"setMaximizable";
	} else if (sel == @selector(isMinimizable)) {
		return @"isMinimizable";
	} else if (sel == @selector(setMinimizable:)) {
		return @"setMinimizable";
	} else if (sel == @selector(isChrome)) {
		return @"isChrome";
	} else if (sel == @selector(setChrome:)) {
		return @"setChrome";
	} else if (sel == @selector(isScrollbars)) {
		return @"isScrollbars";
	} else if (sel == @selector(setScrollbars:)) {
		return @"setScrollbars";
	} else if (sel == @selector(setTransparency:)) {
		return @"setTransparency";
	} else if (sel == @selector(getTransparency)) {
		return @"getTransparency";
	} else if (sel == @selector(isCloseable)) {
		return @"isCloseable";
	} else if (sel == @selector(setCloseable:)) {
		return @"setCloseable";
	} else if (sel == @selector(isVisible)) {
		return @"isVisible";
	} else if (sel == @selector(setVisible:)) {
		return @"setVisible";
	} else if (sel == @selector(setTitle:)) {
		return @"setTitle";
	} else if (sel == @selector(getTitle)) {
		return @"getTitle";
	} else if (sel == @selector(setURL:)) {
		return @"setURL";
	} else if (sel == @selector(getURL)) {
		return @"getURL";
	}
	return nil;
}


+ (BOOL)isKeyExcludedFromWebScript:(const char*)key {
	return YES;
}


+ (NSString *)webScriptNameForKey:(const char *)name {
	return nil;
}

@end

File: runtime/native/win32/src/titanium/ti_window_config.cc
/*
* Copyright 2006-2008 Appcelerator, Inc.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

#include "ti_app_config.h"
#include "ti_window_config.h"
#include "googleurl/src/gurl.h"

int TiWindowConfig::DEFAULT_POSITION = -1;

void TiWindowConfig::setDefaults ()
{
	maximizable = minimizable = closeable = resizable = true;
	usingChrome = usingScrollbars = fullscreen = false;
	transparency = 1.0;
	width = 800;
	height = 600;
	x = y = TiWindowConfig::DEFAULT_POSITION;
	minWidth = minHeight = 0;
	maxWidth = maxHeight = 9000;
	url = "app://" + TiAppConfig::instance()->getAppID() + "/index.html";
	title = "Titanium Application";
	visible = true;
}

std::string TiWindowConfig::insertAppID(std::string url)
{
	GURL gurl(url);
	GURL::Replacements replacements;

	if (gurl.SchemeIs("app")) {
		replacements.SetHostStr(TiAppConfig::instance()->getAppID());
		std::string path = "/" + gurl.host();
		if (gurl.has_path() && gurl.path() != "/") {
			path += gurl.path();
		}

		replacements.SetPathStr(path);

		gurl = gurl.ReplaceComponents(replacements);
	}

	return gurl.spec();
}

TiWindowConfig::TiWindowConfig(xmlElementPtr element)
{
	setDefaults();

	xmlNodePtr child = element->children;
	while (child != NULL) {
		if (nodeNameEquals(child, "id")) {
			id = nodeValue(child);
		}
		else if (nodeNameEquals(child, "title")) {
			title = nodeValue(child);
		}
		else if (nodeNameEquals(child, "url")) {
			url = insertAppID(nodeValue(child));
		}
		else if (nodeNameEquals(child, "maximizable")) {
			maximizable = boolValue(child);
		}
		else if (nodeNameEquals(child, "minimizable")) {
			minimizable =	 boolValue(child);
		}
		else if (nodeNameEquals(child, "closeable")) {
			closeable = boolValue(child);
		}
		else if (nodeNameEquals(child, "resizable")) {
			resizable = boolValue(child);
		}
		else if (nodeNameEquals(child, "fullscreen")) {
			fullscreen = boolValue(child);
		}
		else if (nodeNameEquals(child, "chrome")) {
			usingChrome = boolValue(child);
			const char * scrollbars = (const char *)xmlGetProp(child, (const xmlChar *)"scrollbars");
			if (scrollbars != NULL) {
				usingScrollbars = TiAppConfig::stringToBool(scrollbars);
			}
		}
		else if (nodeNameEquals(child, "transparency")) {
			transparency = (float)atof(nodeValue(child));
		}
		else if (nodeNameEquals(child, "width")) {
			width = atoi(nodeValue(child));
		}
		else if (nodeNameEquals(child, "height")) {
			height = atoi(nodeValue(child));
		}
		else if (nodeNameEquals(child, "min-width")) {
			minWidth = atoi(nodeValue(child));
		}
		else if (nodeNameEquals(child, "max-width")) {
			maxWidth = atoi(nodeValue(child));
		}
		else if (nodeNameEquals(child, "min-height")) {
			minHeight = atoi(nodeValue(child));
		}
		else if (nodeNameEquals(child, "max-height")) {
			maxHeight = atoi(nodeValue(child));
		}
		child = child->next;
	}
}

std::string TiWindowConfig::toString()
{
	std::ostringstream stream;

	stream << "[TiWindowConfig id=" << id
		<< ", x=" << x << ", y=" << y
		<< ", width=" << width << ", height=" << height
		<< ", url=" << url
		<< "]";

	return stream.str();
}